tests/qasm3_qir/converter/test_gates.py

```diff
@@ -222,6 +222,87 @@ def test_custom_ops(test_name, request):
     check_custom_qasm_gate_op(generated_qir, gate_type)
 
 
+def test_pow_gate_modifier():
+    qasm3_string = """
+    OPENQASM 3;
+    include "stdgates.inc";
+    qubit q;
+    inv @ pow(2) @ pow(4) @ h q;
+    pow(-2) @ h q;
+    """
+    result = qasm3_to_qir(qasm3_string)
+    generated_qir = str(result).splitlines()
+    check_attributes(generated_qir, 1, 0)
+    check_single_qubit_gate_op(generated_qir, 10, [0] * 10, "h")
+
+
+def test_inv_gate_modifier():
+    qasm3_string = """
+    OPENQASM 3;
+    include "stdgates.inc";
+    qubit q;
+    inv @ h q;
+    inv @ y q;
+    inv @ rx(0.5) q;
+    inv @ s q;
+
+    qubit[2] q2;
+    inv @ cx q2;
+    inv @ ccx q[0], q2;
+    inv @ u2(0.5, 0.5) q2[0];
+    """
+    result = qasm3_to_qir(qasm3_string)
+    generated_qir = str(result).splitlines()
+    check_attributes(generated_qir, 3, 0)
+    check_single_qubit_gate_op(generated_qir, 1, [0], "h")
+    check_single_qubit_gate_op(generated_qir, 1, [0], "y")
+    check_single_qubit_rotation_op(generated_qir, 1, [0], [-0.5], "rx")
+    check_single_qubit_gate_op(generated_qir, 1, [0], "sdg")
+    check_two_qubit_gate_op(generated_qir, 1, [[1, 2]], "cx")
+    check_three_qubit_gate_op(generated_qir, 1, [[0, 1, 2]], "ccx")
+
+
+def test_nested_gate_modifiers():
+
+    complex_qir = qasm3_to_qir(
+        """
+    OPENQASM 3;
+    include "stdgates.inc";
+    qubit[2] q;
+    gate custom2 p, q{
+        y p;
+        z q;
+    }
+    gate custom p, q {
+        pow(1) @ custom2 p, q;
+    }
+    pow(1) @ inv @ pow(2) @ custom q; 
+    pow(-1) @ custom q;
+    """
+    )
+    generated_qir = str(complex_qir).splitlines()
+    check_attributes(generated_qir, 2, 0)
+    check_single_qubit_gate_op(generated_qir, 2, [0, 0, 0], "y")
+    check_single_qubit_gate_op(generated_qir, 2, [1, 1, 1], "z")
+
+
+def test_unsupported_modifiers():
+    # TO DO : add implementations, but till then we have tests
+    for modifier in ["ctrl", "negctrl"]:
+        with pytest.raises(
+            NotImplementedError,
+            match=r"Controlled modifier gates not yet supported .*",
+        ):
+            _ = qasm3_to_qir(
+                f"""
+                OPENQASM 3;
+                include "stdgates.inc";
+                qubit[2] q;
+                {modifier} @ h q[0], q[1];
+                """
+            )
+
+
 def test_incorrect_custom_ops():
```

Comment:
Let's use `pytest.mark.parametrize` to split this into 5 tests. For example, just for the first two, this would look something like:

```python
import pytest

@pytest.mark.parametrize("qasm_input, error_message", [
    (
        """
        OPENQASM 3;
        include "stdgates.inc";

        qubit[2] q1;
        custom_gate q1;  // undeclared gate
        """,
        r"Unsupported / undeclared QASM operation: custom_gate",
    ),
    (
        """
        OPENQASM 3;
        include "stdgates.inc";

        gate custom_gate(a,b) p, q{
            rx(a) p;
            ry(b) q;
        }

        qubit[2] q1;
        custom_gate(0.5) q1;  // parameter count mismatch
        """,
        r"Parameter count mismatch for gate custom_gate. Expected 2 but got 1 .*",
    ),
])
def test_qasm3_to_qir_errors(qasm_input, error_message):
    with pytest.raises(Qasm3ConversionError, match=expected_message):
        qasm3_to_qir(qasm_input)
```

*Note*: didn't actually test this code so not sure if it works but that's the idea. Will help reduce duplicate code and improve clarity as to which test fails, if any.